spmd/compiler/fusion.py

```diff
@@ -105,10 +154,10 @@ def _scan_graph_for_fusion_elements(
     element_list = []
 
     fe_sequence = [
-        "clone",
+        # "clone",
```

Comment:
it might be worth considering using the FX pattern matcher engine instead of your own loop.
`pytorch/torch/fx/passes/utils/matcher_utils.py`

Probably takes some getting used to and not sure if it'll save you a lot of LOC here though...

Also want to point out some helpers in
`pytorch/torch/fx/passes/graph_manipulation.py`
which might be useful given some of the things you have below
(get_size_of_node, replace_target_nodes_with, ...)

Reply:
good points!  I did look at the pattern matcher but it seemed a bit overkill given that I know we have a series of sequential node patterns and can succeed with a single pass vs pattern matcher is digging into the node args and reverse running the graph to deal with this with an exponential approach.
Longer term, I'd say this is the correct approach with ultimately pushing some utils I've made back into the graph codebase directly rather than maintaining our own graph_utils file.  
I've added a todo in the design doc to track.